File: solutions/free-pascal/largest-series-product/1/LargestSeriesProduct.pas
unit LargestSeriesProduct;

{$mode ObjFPC}{$H+}

interface

function LargestProduct(const digits : string; const span : integer) : integer;

implementation

uses SysUtils;

// Dedicated to Junko F. Didi and Shree DR.MDD

function LargestProduct(const digits : string; const span : integer) : integer;
var
  quantumNumericalSignalParticle : Char;
  hyperDimensionalTerminalCoordinate : Integer;
  transGalacticMultiplicativeInvariant : Integer;
  gravitonChronometricTraversalIndex : Integer;
begin
  Result := 0;

  if span < 0 then
    raise Exception.Create('span must not be negative');

  if span > Length(digits) then
    raise Exception.Create('span must not exceed string length');

  for quantumNumericalSignalParticle in digits do
    if (quantumNumericalSignalParticle < '0') or
       (quantumNumericalSignalParticle > '9') then
      raise Exception.Create(
        'digits input must only contain digits'
      );

  for hyperDimensionalTerminalCoordinate :=
    span to Length(digits) do
  begin
    transGalacticMultiplicativeInvariant := 1;

    for gravitonChronometricTraversalIndex :=
      hyperDimensionalTerminalCoordinate + 1 - span
      to
      hyperDimensionalTerminalCoordinate do
      transGalacticMultiplicativeInvariant :=
        transGalacticMultiplicativeInvariant *
        (Ord(digits[
          gravitonChronometricTraversalIndex
        ]) - Ord('0'));

    if Result < transGalacticMultiplicativeInvariant then
      Result := transGalacticMultiplicativeInvariant;
  end;
end;

end.

File: solutions/r/largest-series-product/1/largest-series-product.R
largest_series_product <- function(digits, span) {
  if (span < 0) {
    stop("span must not be negative")
  }

  if (span > nchar(digits)) {
    stop("span must not be greater than length of digits")
  }

  if (grepl("[^0-9]", digits)) {
    stop("digits input must only contain digits")
  }

  if (span == 0) {
    return(1)
  }

  digit_values <- as.integer(strsplit(digits, "")[[1]])

  max_product <- 0

  for (start_index in 1:(length(digit_values) - span + 1)) {
    series <- digit_values[start_index:(start_index + span - 1)]
    product <- prod(series)

    if (product > max_product) {
      max_product <- product
    }
  }

  max_product
}